File: packages/wasmvm/wasmclient/ts/wasmclient/isc/wasmconvertor.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as isc from "./index";
import * as wasmlib from "wasmlib"

export class WasmConvertor {
    public constructor() {
    }

    public iscAddress(addr: wasmlib.ScAddress): isc.Address {
        return wasmlib.bytesToUint8Array(wasmlib.addressToBytes(addr));
    }

    public iscAllowance(addr: wasmlib.ScAssets): isc.Allowance {
        // TODO
        return [];
    }

    public iscChainID(chainID: wasmlib.ScChainID): isc.ChainID {
        return wasmlib.bytesToUint8Array(wasmlib.chainIDToBytes(chainID));
    }

    public iscHname(hName: wasmlib.ScHname): isc.Hname {
        return wasmlib.uint32FromBytes(wasmlib.hnameToBytes(hName))
    }

    public iscRequestID(chainID: wasmlib.ScRequestID): isc.RequestID {
        return wasmlib.bytesToUint8Array(wasmlib.requestIDToBytes(chainID));
    }

    public scAddress(addr: isc.Address): wasmlib.ScAddress {
        return wasmlib.addressFromBytes(wasmlib.bytesFromUint8Array(addr));
    }

    public scChainID(chainID: isc.ChainID): wasmlib.ScChainID {
        return wasmlib.chainIDFromBytes(wasmlib.bytesFromUint8Array(chainID))
    }

    public scHname(hName: isc.Hname): wasmlib.ScHname {
        return wasmlib.hnameFromBytes(wasmlib.uint32ToBytes(hName));
    }

    public scRequestID(requestID: isc.RequestID): wasmlib.ScRequestID {
        return wasmlib.requestIDFromBytes(wasmlib.bytesFromUint8Array(requestID))
    }
}

File: packages/wasmvm/wasmclient/src/isc/wasmconvertor.rs
pub use crate::types::*;
pub use wasmlib::wasmtypes;

pub struct WasmConvertor {}

impl WasmConvertor {
    pub fn new() -> Self {
        return WasmConvertor {};
    }

    pub fn isc_address(&self, addr: &wasmlib::ScAddress) -> Address {
        return wasmlib::address_to_bytes(&addr) as Address;
    }

    pub fn isc_allowance(&self, addr: &wasmlib::ScAssets) -> Allowance {
        // FIXME
        return addr.to_bytes() as Allowance;
    }

    pub fn isc_chain_id(&self, chain_id: &wasmlib::ScChainID) -> ChainID {
        return wasmlib::chain_id_to_bytes(&chain_id) as ChainID;
    }

    pub fn isc_hname(&self, hname: &wasmlib::ScHname) -> Hname {
        return wasmtypes::uint32_from_bytes(&wasmlib::hname_to_bytes(&hname)) as Hname;
    }

    pub fn isc_request_id(&self, request_id: &wasmlib::ScRequestID) -> RequestID {
        return wasmlib::request_id_to_bytes(&request_id) as RequestID;
    }

    pub fn sc_address(&self, addr: &Address) -> wasmlib::ScAddress {
        return wasmlib::address_from_bytes(addr);
    }

    pub fn sc_chain_id(&self, chain_id: &ChainID) -> wasmlib::ScChainID {
        return wasmlib::chain_id_from_bytes(chain_id);
    }

    pub fn sc_hname(&self, hname: &Hname) -> wasmlib::ScHname {
        return wasmlib::hname_from_bytes(&wasmtypes::uint32_to_bytes(hname.clone()));
    }

    pub fn sc_request_id(&self, request_id: &RequestID) -> wasmlib::ScRequestID {
        return wasmlib::request_id_from_bytes(request_id);
    }
}
